presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java

```diff
@@ -41,12 +44,48 @@
     private final AtomicReference<PrestoConnection> connection;
     private final AtomicReference<ResultSet> currentResult = new AtomicReference<>();
     private final AtomicLong currentUpdateCount = new AtomicLong(-1);
+    private final ProgressCallback progressCallback = new ProgressCallback();
+
+    @ThreadSafe
+    final class ProgressCallback
```

Comment:
That won't work because you will then need to expose the `AtomicReference` to a different class to issue the callback. That breaks encapsulation of thread-safety mechanisms.